### PR TITLE
Start using the checked constructors

### PR DESCRIPTION
**Pull request by edsko**
Currently stuck on thunks inside ChainFragment

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Mock/UTxO.hs [hunk 1]
```diff
@@ -46,14 +50,21 @@ import           Ouroboros.Consensus.Ledger.Mock.Address
...
 
-data Tx = Tx (Set TxIn) [TxOut]
-  deriving (Show, Eq, Ord, Generic, Serialise)
+data Tx = UnsafeTx (Set TxIn) [TxOut]
+  deriving stock    (Show, Eq, Ord, Generic)
+  deriving anyclass (Serialise, NFData)
+  deriving NoUnexpectedThunks via UseIsNormalForm Tx
+
+mkTx :: Set TxIn -> [TxOut] -> Tx
+mkTx ins outs = force tx
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Mock/UTxO.hs]: Why not inline `tx`? Fine either way.

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Mock/UTxO.hs [hunk 2]
```diff
@@ -46,14 +50,21 @@ import           Ouroboros.Consensus.Ledger.Mock.Address
   Basic definitions
 -------------------------------------------------------------------------------}
 
-data Tx = Tx (Set TxIn) [TxOut]
-  deriving (Show, Eq, Ord, Generic, Serialise)
+data Tx = UnsafeTx (Set TxIn) [TxOut]
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Mock/UTxO.hs]: Oh, can't we use a pattern synonym for this? I don't like having to pattern-match on an `UnsafeTx` :grin: 
So keep this constructor, don't export it, but add a bidirectional `Tx` pattern synonym that replaces `mkTx` and is the same as `UnsafeTx` pattern-wise.

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/TxSeq.hs [hunk 3]
```diff
@@ -53,7 +60,14 @@ data TxTicket tx = TxTicket !tx !TicketNo
...
 newtype TxSeq tx = TxSeq (StrictFingerTree TxSeqMeasure (TxTicket tx))
-  deriving Show
+  deriving (Show)
+
+instance NoUnexpectedThunks tx => NoUnexpectedThunks (TxSeq tx) where
+  showTypeOf _ = "TxSeq"
+  whnfNoUnexpectedThunks ctxt (TxSeq ft) =
+      -- We cannot rely on our own 'Foldable' instance here because it
+      -- /introduces/ thunks
+      noUnexpectedThunksInValues ctxt (Foldable.toList ft)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Mempool/TxSeq.hs]: Is there no instance for `StrictFingerTree`?

File: ouroboros-consensus/test-util/Test/Util/TestBlock.hs [hunk 4]
```diff
@@ -114,18 +117,21 @@ import qualified Ouroboros.Consensus.Util.SlotBounded as SB
 -- in-memory representation).
 --
 -- The 'BlockNo' of the corresponding block is just the length of the list.
-newtype TestHash = TestHash {
+newtype TestHash = UnsafeTestHash {
```
Comment (mrBliss) [inline ouroboros-consensus/test-util/Test/Util/TestBlock.hs]: The same pattern synonym trick can be used here.

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Mock/UTxO.hs (hunk 2 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Mock/UTxO.hs]: Done.

File: ouroboros-consensus/test-util/Test/Util/TestBlock.hs (hunk 4 above)
Comment (edsko) [reply ouroboros-consensus/test-util/Test/Util/TestBlock.hs]: Done.

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Mock/UTxO.hs (hunk 1 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Mock/UTxO.hs]: Done.

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/TxSeq.hs (hunk 3 above)
Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Mempool/TxSeq.hs]: There is now..

Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Mempool/TxSeq.hs]: Or rather, we could make one now.

Comment (edsko) [reply ouroboros-consensus/src/Ouroboros/Consensus/Mempool/TxSeq.hs]: Done.

**Pull request merged**